src/main/scala/firrtl/passes/wiring/Wiring.scala

```diff
@@ -0,0 +1,165 @@
+
+    // Populate lineage tree with relationship information, i.e. who is source,
+    //   sink, parent of source, etc.
+    val withFields = setSharedParent(wi.top)(setFields(sinks, source)(lineages))
+
+    // Populate lineage tree with what to instantiate, connect to/from, etc.
+    val withThings = setThings(portNames, compName)(withFields)
+
+    // Create a map from module name to lineage information
+    val map = pointToLineage(withThings)
+
+    // Obtain the source component type
+    val sourceComponentType = getType(c, source, compName)
+
+    // Return new circuit with correct wiring
+    val cx = c.copy(modules = c.modules map onModule(map, sourceComponentType))
+
+    // Replace inserted IR nodes with WIR nodes
+    ToWorkingIR.run(cx)
+  }
+
+  /** Return new module with correct wiring
+    */
+  private def onModule(map: Map[String, Lineage], t: Type)(m: DefModule) = {
+    map.get(m.name) match {
+      case None => m
+      case Some(l) =>
+        val stmts = mutable.ArrayBuffer[Statement]()
+        val ports = mutable.ArrayBuffer[Port]()
+        l.addPort match {
+          case None =>
+          case Some((s, dt)) => dt match {
+            case DecInput => ports += Port(NoInfo, s, Input, t)
+            case DecOutput => ports += Port(NoInfo, s, Output, t)
+            case DecWire =>
+              stmts += DefWire(NoInfo, s, t)
+          }
+        }
+        stmts ++= l.cons.map{case ((l, r)) => 
+          Connect(NoInfo, toExp(l), toExp(r))
+        }
+        def onStmt(s: Statement): Statement = Block(Seq(s) ++ stmts)
+        m match {
+          case Module(i, n, ps, s) => Module(i, n, ps ++ ports, Block(Seq(s) ++ stmts))
+          case ExtModule(i, n, ps) => ExtModule(i, n, ps ++ ports)
+        }
+    }
+  }
+
+  /** Returns the type of the component specified
+    */
+  private def getType(c: Circuit, module: String, comp: String) = {
+    def getRoot(e: Expression): String = e match {
+      case r: Reference => r.name
+      case i: SubIndex => getRoot(i.expr)
+      case a: SubAccess => getRoot(a.expr)
+      case f: SubField => getRoot(f.expr)
+    }
+    val eComp = toExp(comp)
+    val root = getRoot(eComp)
+    var tpe: Option[Type] = None
+    def getType(s: Statement): Statement = s match {
+      case DefRegister(_, n, t, _, _, _) if n == root =>
+        tpe = Some(t)
+        s
+      case DefWire(_, n, t) if n == root =>
+        tpe = Some(t)
+        s
+      case WDefInstance(_, n, m, t) if n == root => 
+        tpe = Some(t)
+        s
+      case DefNode(_, n, e) if n == root =>
+        tpe = Some(e.tpe)
+        s
+      case sx: DefMemory if sx.name == root =>
+        tpe = Some(MemPortUtils.memType(sx))
+        sx
+      case sx => sx map getType
+    }
+    val m =  c.modules.filter(_.name == module) match {
+      case Seq(mod) => mod
+      case _ => error(s"Must have a module named $module")
+    }
+    m.ports.foreach{ p => if(p.name == root) tpe = Some(p.tpe) }
```

Comment:
No, i'm just iterating through the ports to find the one with the correct name. There might be one, there might not be. I could use find, but it would be a bit more verbose i think